app/services/oauth_clients/redirect_uri_from_state_service.rb

```diff
@@ -28,30 +28,46 @@
 
 require "rack/oauth2"
 require "uri/http"
+require 'dry/monads'
+require 'dry/monads/do'
 
 module OAuthClients
   class RedirectUriFromStateService
+    include Dry::Monads[:maybe]
+    include Dry::Monads::Do.for(:process)
+
     def initialize(state:, cookies:)
-      @state = state
+      @state = Maybe(state)
       @cookies = cookies
     end
 
     def call
-      redirect_uri = oauth_state_cookie
-
-      if redirect_uri.present? && ::API::V3::Utilities::PathHelper::ApiV3Path::same_origin?(redirect_uri)
-        ServiceResult.success(result: redirect_uri)
-      else
-        ServiceResult.failure
-      end
+      process(@cookies, @state)
+        .fmap { |uri| ServiceResult.success(result: uri) }
+        .value_or ServiceResult.failure
     end
 
     private
 
-    def oauth_state_cookie
-      return nil if @state.blank?
+    def process(cookies, state)
+      uuid = yield state
+      uri = yield callback_uri_from_cookies cookies, "oauth_state_#{uuid}"
+      callback_uri = yield validate_callback_uri uri
```

Comment:
I am a friend of code that "doesn't make me think". So I prefer parenthesis in most cases.

```suggestion
      callback_uri = yield validate_callback_uri(uri)
```